src/main/java/com/vmware/vipclient/i18n/messages/service/CacheService.java

```diff
@@ -6,127 +6,101 @@
 
 import com.vmware.vipclient.i18n.VIPCfg;
 import com.vmware.vipclient.i18n.base.cache.Cache;
-import com.vmware.vipclient.i18n.base.cache.FormatCacheItem;
+import com.vmware.vipclient.i18n.base.cache.CacheItem;
 import com.vmware.vipclient.i18n.base.cache.MessageCacheItem;
 import com.vmware.vipclient.i18n.messages.dto.MessagesDTO;
 import com.vmware.vipclient.i18n.util.ConstantsKeys;
 import com.vmware.vipclient.i18n.util.LocaleUtility;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
 
 public class CacheService {
     private MessagesDTO dto;
 
     public CacheService(MessagesDTO dto) {
         this.dto = dto;
     }
-    
+
     public MessageCacheItem getCacheOfComponent() {
         String cacheKey = dto.getCompositStrAsCacheKey();
-        Locale matchedLocale = LocaleUtility.pickupLocaleFromList(
-                this.getSupportedLocalesFromCache(),
+        Locale matchedLocale = LocaleUtility.pickupLocaleFromList(this.getLocalesOfCachedMsgs(),
                 this.getLocaleByCachedKey(cacheKey));
         cacheKey = cacheKey.substring(0,
                 cacheKey.indexOf(ConstantsKeys.UNDERLINE_POUND) + 2)
                 + matchedLocale.toLanguageTag();
-        Cache c = VIPCfg.getInstance().getCacheManager().getCache(VIPCfg.CACHE_L3);
-        if (c == null) {
-            return null;
-        } else {
-            return (MessageCacheItem) c.get(cacheKey);
-        }
+        return (MessageCacheItem) this.getCacheItem(cacheKey);
     }
 
     public void addCacheOfComponent(MessageCacheItem itemToCache) {
         String cacheKey = dto.getCompositStrAsCacheKey();
-        Cache c = VIPCfg.getInstance().getCacheManager().getCache(VIPCfg.CACHE_L3);
-        if (c != null) {
-            c.put(cacheKey, itemToCache);
-        }
+        this.addCacheItem(cacheKey, itemToCache);
     }
 
     public void updateCacheOfComponent(MessageCacheItem itemToCache) {
         String cacheKey = dto.getCompositStrAsCacheKey();
-        Cache c = VIPCfg.getInstance().getCacheManager().getCache(VIPCfg.CACHE_L3);
-        if (c != null) {
-            MessageCacheItem cacheItem = (MessageCacheItem) c.get(cacheKey);
-            if (cacheItem == null) {
-            	cacheItem = new MessageCacheItem();
-            	c.put(cacheKey, cacheItem);
-            }
-            cacheItem.setCacheItem(itemToCache);
+        MessageCacheItem cacheItem = (MessageCacheItem) this.getCacheItem(cacheKey);
+        if (cacheItem == null) {
+            cacheItem = new MessageCacheItem();
+            this.addCacheItem(cacheKey, cacheItem);
         }
-    }
-
-    public boolean isContainComponent() {
```

Comment:
Please don't remove this method, this is to fix a bug about source collection. Need to store null objects to cache to avoid bugs. 
You can confirm this with Colin.

Reply:
I put it back, but if you check the code, it is exactly the same as `CacheService.getCacheOfComponent()!=null`

On that note, we are not going to use the java client library for source collection anymore, even for existing products.
Also, if you check where CacheService.addCacheOfComponent is invoked, you will see that we do not store null in the cache. We do not even store a MessageCacheItem with an empty map.

Reply:
@Xiaochao8 , did you check where this method is used? Do you agree that we can remove it to simplify code?